lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+        """Set the ca_certificate on the relation.
+
+        Args:
+            relation: The relation for which to set the ca_certificate.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if not relation:
+            logger.warning("Relation is None")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        relation.data[self.charm.app]["ca_certificate"] = ca_certificate
+
+
+def is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Return whether the provider data is valid.
+
+    This uses the pydantic schema to validate the data.
+
+    Args:
+        data: The data to validate.
+
+    Returns:
+        True if the data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        if event.app is None:
+            logger.warning("No remote application yet")
+            return
+        if is_provider_data_valid(event.relation.data[event.app]):
+            self.on.vault_autounseal_details_ready.emit(event.relation)
+
+    def get_vault_url(self, relation: ops.Relation) -> Optional[str]:
+        """Return the vault url from the relation databag.
+
+        Args:
+            relation: The relation from which to get the vault url.
+
+        Returns:
+            The vault url if it exists, None otherwise.
+        """
+        if relation.app is None:
+            return None
+        return relation.data[relation.app].get("address")
+
+    @staticmethod
+    def get_credentials_secret_id(relation: ops.Relation) -> Optional[str]:
+        """Return the credentials secret id from the relation.
+
+        Args:
+            relation: The relation from which to get the credentials secret id.
+
+        Returns:
+            The credentials secret id if it exists, None otherwise.
+        """
+        if relation.app is None:
+            return None
+        return relation.data[relation.app].get("credentials_secret_id")
+
+    def get_credentials(self, credentials_secret_id: str) -> tuple[str | None, str | None]:
```

Comment:
Does this need to be public? We expect the requirer to interface with `get_credentials_from_relation` right?

Reply:
I don't think it really matters. Does it need to be private? I don't see any point in restricting people not to use this (and also, vault_kv of course).

Reply:
no longer relevant in https://github.com/canonical/vault-k8s-operator/pull/346/commits/0651467cd3a75065ba0f04391194ceef96844526